### DISCUSSION TITLE
warning: loop not unrolled: the optimizer was unable to perform the requested transformation;

### DISCUSSION THREAD

**Discussion by zjin-lcf** [intel/llvm#4086] [General]
Is unrolling (e.g #pragma unroll 32) disabled in a SYCL kernel ?  
Can a user force loop unrolling ?  

Thanks

**Comment by bader** [comment]
It's disabled in DPC++ front-end, but back-end compiler should be able to unroll the loop.
Do you see the loop in the generated assembly code?

**Reply by zjin-lcf** [reply]
Should the front-end and back-end produce consistent message about loop unrolling ? 

Are there ways to know that a loop is unrolled without looking at the generated assembly ? 

Thanks

**Reply by bader** [reply]
> Should the front-end and back-end produce consistent message about loop unrolling ?

Yes, they should and they do as far as I know. At some point there was a bug in the front-end compiler, but it's fixed already. Which compiler version do you use? If you see that message from the latest nightly build it's either a new bug (if loop is unrolled) or there are legal reasons to apply the transformation.

> Are there ways to know that a loop is unrolled without looking at the generated assembly ?

AFAIK, assembly is the most reliable way to validate compiler behavior. OTOH, LLVM optimizations has some mechanisms to produce optimization reports. Probably you should just file an issue and attach a reproducer for that issue. According to my understanding #pragma unroll is not disabled in a SYCL kernel.

Tagging @xtian-github, @andykaylor and @hidekisaito for more insights.

**Reply by zjin-lcf** [reply]
I was not building a SYCL program with the latest version, and will try new versions.  I will try the optimization report when the compiler has some command-line options for producing a report. Thanks.
